File: 589 - N-ary Tree Preorder Traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    res = []
    def preorder(self, root: 'Node') -> List[int]:
        self.res.clear()
        self.dfs(root)
        return self.res
    def dfs(self, root):
        if root is None:
            return
        self.res.append(root.val)
        for child in root.children:
            self.dfs(child)
class Solution:
    res = []
    queue = []
    def preorder(self, root: 'Node') -> List[int]:
        self.res.clear()
        self.dfs(root)
        return self.res
    def dfs(self, root):
        if root is None:
            return
        self.queue.append(root)
        while len(self.queue):
            t = self.queue.pop()
            self.res.append(t.val)
            for child in t.children[::-1]:
                self.queue.append(child)

File: 589 - N-ary Tree Preorder Traversal.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node} root
 * @return {number[]}
 */
let res = [];
var preorder = function (root) {
    res = [];
    dfs(root);
    return res;
};
function dfs(root) {
    if (root == null) {
        return;
    }
    res.push(root.val);
    for (let child of root.children) {
        dfs(child);
    }
}
let a = {
    value: 1
};
let b = {
    value: 1
};
let c = [1, 'a', {}];
let d = [1, 'a', {}];
console.log(c === d);
console.log(typeof (6 * '2'), typeof '12');
